### ISSUE TITLE
Failure sending multiple simultaneous emails

### ISSUE THREAD

**Issue by atuzovic** [jstedfast/MailKit#626]
After upgrading our solution from MailKit 1.x version to 2.0.1 version, we have started to experience errors whenever we attempt to send e-mails on multiple threads at the same time.

We receive one of the two following errors:

> Error reading a response from the SMTP server. => Cannot access a disposed object.
> Object name: 'SslStream'.

> Error reading a response from the SMTP server. =>  The BeginRead method cannot be called when another read operation is pending.

Is there any reference on how to properly use new asynchronous APIs?

This is code sample from our app:

    public class MailerProcess
    {
        public static void Run()
        {
             using (SmtpClient client = new SmtpClient())
             {
                   string host = AppSettingsWrapper.Host;
                   string port = AppSettingsWrapper.Port.ToString();

                   client.Connect(AppSettingsWrapper.Host, AppSettingsWrapper.Port, SecureSocketOptions.Auto);
                   client.Authenticate(AppSettingsWrapper.UserName, AppSettingsWrapper.Password);
                   client.Timeout = AppSettingsWrapper.Timeout;

 	           List<Task> taskList = new List<Task>();
 
                   foreach (MailerJob job in batch)
                   {
                        taskList.Add(SendAsync(client, fm, job.Email));
                   }

                   Task.WhenAll(taskList).Wait();

                   if(client.IsConnected)
                   {
                        client.Disconnect(true);
                   }
              }
        }

        private static async Task SendAsync(SmtpClient client, IFileManager fm, Email email)
        {
            try
            {
                MimeMessage message = new MimeMessage();

                MailboxAddress from = new MailboxAddress(email.Content.FromDisplayName, email.Content.FromAddress);
                MailboxAddress to = new MailboxAddress(email.Content.ToDisplayName, email.Content.ToAddress);
                message.From.Add(from);
                message.To.Add(to);

                message.Subject = email.Content.Subject;

                BodyBuilder bodyBuilder = new BodyBuilder();
                if (email.Content.IsHtmlContent)
                {
                    bodyBuilder.HtmlBody = email.Content.Body;
                }
                else
                {
                    bodyBuilder.TextBody = email.Content.Body;
                }

                message.Body = bodyBuilder.ToMessageBody();

                if (client.IsConnected)
                {
                    await client.SendAsync(message);
                }
                return;
            }
            catch (SmtpCommandException se)
            {
                email.StatusMessage = se.Message;
                email.StatusCode = (int)se.StatusCode;
            }
            catch (Exception e)
            {
                email.StatusMessage = e.RecursiveMessage();
            }
        }
    }

Run method is invoked from .NET Core 2.0 console application.

**Comment by jstedfast** [comment]
What you want to do is use:

```csharp
lock (client.SyncRoot) {
     await client.SendAsync (message);
}
```

Hope that helps.

**Comment by atuzovic** [comment]
@jstedfast - Thanks for the quick reply.

Isn't this preventing code from executing in parallel?

**Comment by jstedfast** [comment]
Yes, that's exactly the point of it :)

You can't send multiple messages over the same socket in parallel.

**Comment by atuzovic** [comment]
@jstedfast Ok, thanks. :)

So how can we use multiple sockets? Create new instance of SmtpClient on every thread?

**Comment by jstedfast** [comment]
Correct.

**Comment by NickSpag** [comment]
Looks like calling await in a lock is illegal? I'm running in to this while trying to optimize a Xamarin.Mac client's full inbox sync. 

I have parallel Tasks (think Task.WhenAll/Any) that do the full processing of a message, and I'm having a hard time determining if the best route is using a shared Semaphore or something like it for the core GetStreamAsync call of each individual message request, or if I should create a handful of ImapClients that each handle a batch of requests synchronously, or a combination of the two. Is the overhead of connecting/authenticating an ImapClient connection worth something like that? 

Apologies if this should be its own issue in stackoverflow or some other channel. Thank you for the library!
